ISLANDORA-2058 Redirect to first item of a compound when going to a compound

lucasvanschaik opened the pull request:
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2058

https://groups.google.com/forum/m/#!topic/islandora/ue4SMkvhdp4

# What does this Pull Request do?

Adds a configuration option that (if enabled) always redirects to the first item of the compound when a compound is viewed. The manage tab of the compound is still reachable.

# What's new?
When a compound is opened, it looks like the first item of that compound is selected. But the context on the page is that of the compound, causing other blocks on the page to display the information about the compound instead of the first item. For our users this is confusing and hard to explain. We can tell them to always click the first item, but we rather redirect to the first item automatically.
In this PR the hook_page_build is implemented to redirect to the first child if the current object has children. Also, in the admin form an option is added the enabled or disable this behaviour. If it is enabled and also the "display breadcrumbs" option is enabled, a warning is displayed because these two don't work together well.

# How should this be tested?

* click on a link to a compound
  * the actual compound is visible and you can see that the first item is displayed but not actually selected
* enable the redirect to first option in the admin form of the compound sp
* click on a link to a compound
  * now the first item of the compound is displayed and selected
* test with the JAIL option enabled
* test with Compound Member Query set to "Legacy SPARQL" and "SPARQL"
* test with other objects as compounds; "Only allow compound objects to have child objects" is disabled
* test with child with 2 parents; this gives unexpected behaviour, but is the same as when a user clicks on the first child of a compound and then finds that another compound is displayed. See JIRA ISLANDORA-1301.

# Additional Notes:
Is related to JIRA ISLANDORA-1301 but does not solve it.
The documentation should be updated, is not done because of screenshot.
This does not add any dependencies and can be enabled/disabled at any time.

# Interested parties
@Islandora/7-x-1-x-committers @whikloj @DiegoPino

lucasvanschaik committed:
Redirect to first child

When the new setting "islandora_compound_object_redirect_to_first" is
TRUE and the current URL points to a compound object with children,
then a redirect is generated to go to the first child.
This solves the problem that blocks on compound views get the wrong
context, because without the redirect the current object is the
compound but the data of the first child is displayed.

lucasvanschaik committed:
Validate redirect to first.

Warn the user if both 'Redirect to first child when a compound object
is requested' and 'Display compound object parents in the breadcrumbs
on children objects' are enabled because this may give unexpected
behaviour.

lucasvanschaik committed:
Improve code.

- follow coding conventions
- make code more readable
- do redirect 302 instead of 301.

DiegoPino commented:
@lucasvanschaik this is great and congratulations on your first official pull request to Islandora 🎈 
Will give it a spin during the day. Also, nice to see a pull request pass Drupal Coding standards at first. Good work. @whikloj will have for sure comments on this too (as maintainer and user).

whikloj commented on islandora_compound_object.module:
```diff
@@ -261,6 +261,33 @@ function islandora_compound_object_islandora_compoundCModel_islandora_view_objec
 /**
+ * Implements hook_page_build().
+ */
+function islandora_compound_object_page_build(&$page) {
```
Did you try adding your logic to the above `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` function?

I'm wondering if adding your check and redirect there wouldn't make more sense. I have to test, but I think this way it will run both functions which means 2 calls to `islandora_compound_object_get_parts()`. For performance reasons I think it would be beneficial to combine them.

I'll try and test this tomorrow.

lucasvanschaik replied:
I use `hook_page_build()` because this hook is called early and not much unneeded work is done when the redirect is actually needed. I've also looked at `hook_boot()` (but that is too early) and `hook_init()` (but this is not called on cached pages). Also, project conditional_redirect uses the same hook.
Is `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` not only called voor compound Cmodel objects? Will this work if "Only allow compound objects to have child objects associated with them" is disabled?
For performance reasons, isn't it better to implement a (single page request) cache in `islandora_compound_object_get_parts()` if it is expensive and it might be called multiple times? (isn't it called twice when JAIL is enabled?)

whikloj commented:
@lucasvanschaik just so I am understanding your use case. It doesn't sound like you are using compound objects. Could you describe a simple example object so I can be sure to test using your use case as well as my own.

lucasvanschaik commented:
We do use compound model objects, but don't need to have the actual compound displayed, because what we want to display, is displayed with every child of the compound. We always display the metadata of the parent compound object when we display a child. Some examples (work in progress):
https://digitalcollections.universiteitleiden.nl/view/item/54344
https://digitalcollections.universiteitleiden.nl/islandora/object/item%3A230457
https://digitalcollections.universiteitleiden.nl/view/item/52994
https://digitalcollections.universiteitleiden.nl/view/item/51749

DiegoPino commented:
Lucas. I'm away today ( on transit on public transport) but had a chance to
test yesterday. I like to hook used to catch the request because it is more
flexible and follows our current trend of using things more the Drupal way.
I have a few minor comments on performance and wording used to expose the
options
In the admin form. I also think we could use conditional form elemeny
states to hide the conflicting option mix additionally to the element
validation you already did. Thanks!

El El vie, 8 de sep. de 2017 a las 11:19, Lucas van Schaik <
notifications@github.com> escribió:

> We do use compound model objects, but don't need to have the actual
> compound displayed, because what we want to display, is displayed with
> every child of the compound. We always display the metadata of the parent
> compound object when we display a child. Some examples (work in progress):
> https://digitalcollections.universiteitleiden.nl/view/item/54344
>
> https://digitalcollections.universiteitleiden.nl/islandora/object/item%3A230457
> https://digitalcollections.universiteitleiden.nl/view/item/52994
> https://digitalcollections.universiteitleiden.nl/view/item/51749
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_solution_pack_compound/pull/130#issuecomment-328133082>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn85z6hAzrX3iEd3_v1zbSnt53XC3OBks5sgVsSgaJpZM4PP5yr>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

whikloj commented:
Ok, so here are my concerns
1. the `hook_page_build` runs twice when clicking on a compound object, not sure why.
1. The `hook_islandora_CMODEL_PID_islandora_view_object()` already does the work of limiting code to the desired content model (compound) in this case.
1. I built a compound with images from a different collection, so clicking on the compound took me away from collection (I know this is dependant on a specific manner of building your compound).

On the plus side it does seem to run before the `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` does so we aren't running that code.

@DiegoPino while I do like being more Drupaly in CLAW, here in 7.x with the existing code infrastructure it seems we would be tacking on additional ways of doing the same thing. This increases the complexity of the code and unless we are moving or planning to move away from the existing hooks then we end up with multiple places to look to debug and test. 

All of this being said, I think the use case is valid. I'm just wondering about something like [**this**](https://github.com/LeidenUniversityLibrary/islandora_solution_pack_compound/compare/7.x-ISLANDORA-2058...uml-digitalinitiatives:7.x-ISLANDORA-2058)

whikloj commented:
Oh, @lucasvanschaik you'll need to add your new variable below [here](https://github.com/Islandora/islandora_solution_pack_compound/blob/7.x/islandora_compound_object.install#L33) so it gets removed on uninstall.

DiegoPino commented:
@whikloj the reason I think the Drupal way applies correctly here is that on redirect we really want to avoid the extra processing logic that islandora hooks apply at all. Also, there is already a module in Drupal that does it like that which is a good practice if we mimic it. We do have some Drupal like approaches Vanilla Islandora. I agree on the need to remove the variable on uninstall and some wording fixes. 
```hook_islandora_CMODEL_PID_islandora_view_object``` has always been confusing seen we allow not-compounds (and that is not a hack) to be compoundable. I think this optional feature adds usability to a very tricky UX. 
I feel we could use here some performance improvements also because the get children and get parent (sparql) is sooo slow. =)

whikloj commented:
@DiegoPino I guess my point is that we are exchanging the extra processing of islandora hooks on each islandora object for the extra processing of these menu_get_item and menu_get_object calls on each page build.

I don't know which costs/saves more, but I know that this approach adds a new layer and the complexity it brings.

But if you can see the benefits, then I will defer as I am certain there is much in Drupal I don't fully understand.

lucasvanschaik committed:
Delete variable on uninstall.

Cleanup the variable islandora_compound_object_redirect_to_first when
the module is uninstalled.

DiegoPino commented:
@whikloj A) are all your concerns addressed and B) had you a change to[ test this ]?(https://github.com/Islandora/islandora_solution_pack_compound/pull/130#discussion_r137590078)
 @lucasvanschaik what is the reason  "display breadcrumbs" won't work? Could we for example alter the breadcrumb array (there is a hook for that) in case this is enabled? In my opinion that is my only concern right now since removing breadcrumb could be _not_ an option for many. Thanks!

whikloj commented:
@DiegoPino as I said my concerns are that we are adding complexity to the creation of a compound to get this interaction which can be achieved with much less code. But I have not done any performance testing to see what the relative benefits of each are. 

As I said before
> But if you can see the benefits, then I will defer as I am certain there is much in Drupal I don't fully understand.

DiegoPino commented:
@whikloj sorry, I don't get how this adds complexity to the "creation". Probably I got that wrong there from your comment. It is true that we don't use that page hook in islandora right now, as we don't use many others, but intercepting the call before the whole islandora calling hooks logic gets in place seems perfomant logic to me. Also ```menu_get_item```is something we use right now in openseadragon module, so that way is not new to us. Othe contributed, but not vanilla ones also use it. I also undertstand why menu_get_item is called and not menu_get_object. In this case we still don't now what URL/menu item is being hit when the page hook gets the path.

I don't want to impose this mechaninc, just being supportive of an idea that looks ok to me and again, if you are not convinced i would prefer concensus and agreement (and that means me working for that). Will bring this to today's committers call and see what others think. Thanks for reviewing.

lucasvanschaik commented:
@DiegoPino the breadcrumbs will work but maybe not as intended; this is because the "compound object" in the breadcrumb is redirected to the first item. So if a user is already on the first item the breadcrumb is something like: a compound object (that redirects to the first item) > the first item.
This might be confusing, so that is the warning for. I can investigate if the breadcrumb array can be altered, but this makes it even more complex...

DiegoPino commented:
Before going for that let me ask committers today at the call we have every two weeks and see what other concerns, ideas, are around this. That said,  if that is the only issue with the breadcrumb i don't see an incompatibility at all. You are simply redirected to an item and that item is marked as the current. Of course if people pinch on the actual compound, parent breadcrumb item, they will go back to the first item, which could be the current page, but if you are already looking at child number 510, and you click, being able to jump to the start again is actually nice... and being very sincere, the whole navigation was weird before( i had the issue of not being quite clear where I was (digitally) actually every time I used compound, first children or the actual compound?). If you want to alter it it goes like this
https://drupal.stackexchange.com/questions/69363/altering-the-breadcrumb-contents

So. I remember asking you this in some email, but how do you administer the compound object itself if you are being redirected? can't remember

lucasvanschaik commented:
The redirect is only done on pages with a path like islandora/object/%. So, the manage parent url (islandora/object/%/manage) is not redirected, it still works and is available on a compound page.

bryjbrown commented:
Hi @lucasvanschaik! First and foremost, I want to thank you for your contribution. Your code does indeed work and solve the intended problem, which is always a cause for celebration. I think the concern here (as I see it anyways) is whether this redirect code should be implemented in a new `islandora_compound_object_page_build()` function, or in the existing `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` function.

Since both options would work and have slightly different performance effects, this is entirely a matter of opinion based on what is most important to the individual. Is it pure performance, readability, adherence to the Drupal Way, or conformance the overall style used in other Islandora modules?

In my humblest of opinions, as someone who is not a Drupal pro but commonly works in Islandora code, if I were to try and debug a potential future issue with redirection and I saw an existing `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` function, that would be the first place I would look since these hooks were created for this specific use case (wanting code that only fires on certain cmodel page loads). @whikloj's earlier example using `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` also seemed to require less lines of code, making it a bit more readable and less complex, which is always appreciated in a codebase.

Additionally, if my understanding how the hooks would fire is correct, it sounds like `islandora_compound_object_page_build()` may be faster overall, but fire on every page build, while  `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` would be slower, but only fire on compound cmodel page builds. If this is correct, then I would have to favor `islandora_compound_object_islandora_compoundCModel_islandora_view_object()` again, since I'd want to stay away from anything that slows down the whole system vs a small slice of the system.

I'll mirror what @whikloj said earlier though, in that there is lots about Drupal that I don't know, and like I said earlier, this is really more of a preference than the code itself being truly problematic. I don't feel strongly enough about which is "right" to further obstruct the process. The most important thing is that the code gets merged before @lucasvanschaik gets soured on wanting to make future contributions! :+1:

lucasvanschaik commented:
There has been alot of discussion, more than I thought was needed for this (relatively) simple change. But this is good because it made me see things from different perspectives and rethink my code.
Because there are concerns about using `hook_page_build` over performance/readability/style, I want to address these points: 
1) I don't know enough about doing it the Drupal or Islandora way, so I don't have a clear opinion about that. 
2) Doing the redirect in `hook_islandora_cmodel_islandora_view_object` is more readable and requires less code, so I see that point. 
3) The performance can be tested, so I did that. I did not use a real performance measuring tool, but just enable the `page timer` of the develop module and reloaded several pages 10 times. To my surprise the `hook_page_build` was significant slower than `hook_islandora_cmodel_islandora_view_object`. Not only on compound object pages, but on every object page, because (as @whikloj already said) the `hook_page_build` is called for every page.

Conclusion: I will change my code (give me a few days) and put the redirect in `hook_islandora_cmodel_islandora_view_object`, because it has better performance and is more readable.

Other conclusion: `islandora_compound_object_get_parts` is slow and is called for every page (because the navigation block is on every islandora object page, `hook_block_view` is called, which calls `islandora_compound_object_navigation_block`, that calls `islandora_compound_object_retrieve_compound_info` and that calls `islandora_compound_object_get_parts`). Making `islandora_compound_object_get_parts` faster would benefit every load of the object page.

lucasvanschaik committed:
Move redirect logic

After considering all comments/suggestions, the redirect logic is moved
to the hook_islandora_cmodel_islandora_view_object function.
This is both more readable and has better performance.

DiegoPino commented:
Thanks @lucasvanschaik 
islandora_compound_object_get_parts is slow: true.
If your findings show this is faster, perfect.
If this is bound to the compoundCmodel view hook then:
1. the admin form setting toggle needs documentation (a note) stating that redirect works only for compounds that have actually that CMODEL on their top objects, since this SP allows also non compound CMODEL to act as compounds/aggregating objects
2. Same goes for the readme (not needed here as part of this pull request but leave that decision to the reviewers and we need to add a note somewhere

whikloj commented:
@lucasvanschaik thanks for doing some testing on this. I agree that the `get_parts()` function is slow. I haven't spent much time digging into it, but that could certainly be a next angle of attack to improve compound display performance. 👍 

@DiegoPino I haven't tested using non-compounds as parents, but I think the display would mimic this functionality anyways.

For example: a compound is (usually) nothing but an empty shell for combining the parts of the compound, so redirecting to the first object makes sense.

But if you use a `large_image` and add child objects, then the first object should be the `large_image` itself so you would actually see that object. Which is the same as this new redirect.

All that is to say I wouldn't worry too much about making a huge distinction that this toggle would not affect non-compound compounds. You are correct, but those objects should already operate as if this toggle was in place.

DiegoPino commented:
@whikloj, I know i can be picky, sorry 😞 But consistency is the key for UI (or documentation that clarifies inconsistency) 

> I think the display would mimic this functionality anyways.

If I remember correctly the display will stay on the actual parent (whatever it is) which is not the same as the first child.
 As you correctly stated, if the parent(the "compounding" object) is an image, it will show that object first(which is not the first child but since it has metadata that makes sense).
 Sadly one thing that changes radically is how you manage the actual parent/compounding object 

When redirecting/(this option on) you have to actually press "manage parent", but when that is enabled for noncompounds you manage it via "manage" tab directly. So there are differences, subtle, but enough to make me a picky human being.
I get it mimics the desired functionality, still adds some little exceptions I would prefer we could comment. I'm not good at wording this stuff into forms so of course it can be better/different. Thanks

bondjimbond commented:
@DiegoPino @whikloj Forgive my ignorance, but... Based on the last couple of comments here... If a normal object like a Large Image can also be a compound parent, why do we have a separate Compound Object type in the first place? Is this another one of those legacy behaviour things?

DiegoPino commented:
@bondjimbond sadly, out of the scope of this pull. There was a time where all was allowed and people had fun doing so. That comes from that time for sure? You can do a blame and find when that was introduced.

jonathangreen commented:
@bondjimbond @DiegoPino I think some of it comes from a debate at one point as to weather it was good idea or not to have objects have multiple content models, or if items should have only one content model....

lucasvanschaik commented:
If I need to make other changes to this PR (textual or otherwise), please let me know. The messages in the UI should be made more understandable by a native speaker.

DiegoPino commented on includes/admin.form.inc:
```diff
@@ -94,6 +94,14 @@ function islandora_compound_object_admin_form($form, &$form_state) {
+  $form['islandora_compound_object_redirect_to_first'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Redirect to first child when a compound object is requested.'),
+    '#description' => t('Compound objects themselves redirect to there first child when this setting is enabled. This solves the problem that blocks on compound views get the wrong context.'),
```
I would suggest
` t('Compound objects redirect to their first child when this setting is enabled') `
and I would remove the reason for this.
OR

 `t ('Users will be redirected to a Compound Object's first child when enabling this setting.'),`

OR

`t ('Users will be redirected to the first child of a Compound Object when enabling this setting.'),`

OR

@whikloj @jonathangreen you are native English speakers. Advice/corrections?

whikloj replied:
I like option C
`t ('Users will be redirected to the first child of a Compound Object when enabling this setting.'),`

bryjbrown replied:
Not to complicate things here even further, but the standard practice on these checkboxes is for the label to be a statement indicating what happens when the checkbox is checked. For instance, in Scholar the checkbox labels are "Enable RoMEO attempts", "Render Google Scholar Search link", "Let users choose display CSL", etc. 

I would recommend having this type of statement for the checkbox, for example "Redirect to first child of Compound Objects" or similar.

lucasvanschaik replied:
@DiegoPino @whikloj Should I change the title of the checkbox to `t ('Users will be redirected to the first child of a Compound Object when enabling this setting.'),` ? Leave the description or remove it? Any other things I should change?
I have a feeling this is an ongoing discussion where someone needs to make a decision.
I don't mind changing the PR, but like to reduce the number of times that I need to do that. So if I know exactly what to do, I will add it in another commit. Please let me know.

whikloj replied:
@lucasvanschaik Change the `#description` of the checkbox to `t('Users will be redirected to the first child of a Compound Object when enabling this setting.')`

whikloj replied:
and sorry for the delay in responding :man_facepalming:

DiegoPino commented on includes/admin.form.inc:
```diff
@@ -126,6 +134,28 @@ function islandora_compound_object_admin_form_jail_validation(array $element, ar
+        && $form_state['values']['islandora_compound_object_show_compound_parents_in_breadcrumbs'] == 1) {
+      $first_title = rtrim($element['#title'], '.!?');
+      $parent_title = rtrim($form['islandora_compound_object_show_compound_parents_in_breadcrumbs']['#title'], '.!?');
+      drupal_set_message(t("Enabling both %first and %parent is not recommended, because the compound object parent in the breadcrumb will be redirect to the first child, so cannot be actually reached.",
```
Some advice here @whikloj @jonathangreen (as native English representatives) 
I feel we need to be should use just one form to refer to the actual parent. The compound object, the compound object parent, or any other. So if using compound object's parent here, we should maybe update  it also at https://github.com/Islandora/islandora_solution_pack_compound/pull/130/files#diff-0cdd19abe14e51fe9e7f061ec72c86d3R100
Also, I still don't see harm on using this together with  "Display compound object parents" on the breadcrumb, maybe this should be just an advice type (Warning as defined now is ok  maybe?) . Mostly because even without actual parent access/ metadata, the breadcrumb shows the actual title of the parent, which in many cases defines semantic context!

Also, i really would like to see a warning/as this one if
```PHP
$form['islandora_compound_object_compound_children'] = array(
     '#type' => 'checkbox',
     '#title' => t('Only allow compound objects to have child objects associated with them'),
     '#description' => t('If unchecked, all objects may have child objects.'),
     '#default_value' => variable_get('islandora_compound_object_compound_children', TRUE),
   );
````
 is enabled and redirect also. Because even if the functionality will be close, it will not be the same. No redirect on those objects, different managment of the parent, different behaviour of breadcrumb. Could that be done?
@whikloj @jonathangreen could you give us a hand here with the wording?

whikloj replied:
@DiegoPino I haven't had a chance to try this out again, so I'll need to find time to do that before I can say.

As for the second concern, I still need to test but I think that if you use a large_image as the parent of your compound, then you get that object first. Including all metadata. Which is the same as you get by using a compoundCModel object with a large_image as the first child with this new toggle enabled.

Managing the parent is the same for both and the redirect states `Redirect to first child when a compound object is requested`. Which is not the same as requesting a large_image object with children 😉

DiegoPino replied:
@whikloj thanks a lot. When you have some time please give my second concern a test. At least the management part (where to click) is actually not the same for the parent. But you are right, if I'm alert (morning coffee) and actually read that sentence calmly it is clear that it won't redirect if the parent is a newspaper. 🎉  Could be even 2xClearer if it says `"Redirect to the first child when an object of type compoundCMODEL is requested." or similar in actual good English. Actually, you are right. That piece should be ok. Still give it a spin when you can. Thanks

DiegoPino reviewed (commented):
Last wording comments.

DiegoPino commented:
@lucasvanschaik code looks good. Added a few comments to the wording/warning/messages part. Not a native English speaker as you already noticed. Pinged @whikloj @jonathangreen and @bryjbrown ) (almost the whole TAG :smile:) for wording advice there. I just would like (and it's just a desire, the reason why I did not make my change requests mandatory) a brief explanation/message or warning when non compound Content Model compounding and this option are enabled together. That would make me happy. Once at least the current wording part is solved we will be good to approve and merge. Took longer than expected. It is not always the case, but in such an already established software with well-known and established expected behavior, even when weird sometimes, new behavior requires sometimes more time to be explored, tested and accepted. Sorry for that. Thanks a lot for your hard work on this.

whikloj commented:
@DiegoPino I wouldn't worry to much about the text around non-compound compounds. I just built a basic image with 2 other images added to it. 

It doesn't really work at all.

When you first click the image from the collection you get the original image, but the compound navigation block only shows the 2 children so you can't get back to the first image unless you go to the collection and back in.

Perhaps this is how a non-compound compound is supposed to work. But I would never suggest anyone to use it.

lucasvanschaik committed:
Textual change suggested by @whikloj

whikloj commented:
@DiegoPino any other concerns here, otherwise I think we are good to merge this. If you agree, then please press the button. 👍

whikloj commented:
@lucasvanschaik thanks for hanging in there for all the discussion and your contribution. 😃

DiegoPino commented:
@lucasvanschaik @whikloj done, merged and seems everybody's concerns addressed. Good, good. Nice contribution
